### PR TITLE
Issue #9854: ModiferOrderCheck should support 'sealed' and 'non-sealed'

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #9854: ModiferOrderCheck should support 'sealed' and 'non-sealed'

Diff Regression projects: https://gist.githubusercontent.com/nmancus1/79c9d575d541725cd0b861497d2e7aa1/raw/38c5c2e22dc7809dcbd89c488bf37f39914b6e15/projects.properties

Diff Regression config: https://gist.githubusercontent.com/nmancus1/b111673a252cfdfa990b562361608e29/raw/344468c69278dbff8c8834a43db19f5c6a7b6eb5/config.xml

### CODE REVIEW

**Comment by nmancus1** [comment]
Github, generate report

**Comment by nmancus1** [comment]
Github, generate report

**Review by AkMo3 — CHANGES_REQUESTED** [review]
> Permitted classes can be final , sealed or non-sealed.

From [java-15-features-sealed-and-hidden-classes-pattern-matching-and-text-blocks](https://betterprogramming.pub/java-15-features-sealed-and-hidden-classes-pattern-matching-and-text-blocks-38f4efdc8adc)
_________________________________________________________________________
Is it possible to add an example of the `permits` class being `final` and `non-sealed` so that every case can be covered?

**Comment by nmancus1** [comment]
> `final` and `non-sealed`

Just to be clear, can you provide an example?

If you mean a permitted class/ interface that is BOTH `non-sealed` and `final`, These modifiers are not allowed to be used together:

`final` = class that cannot be extended
`non-sealed` = a way to "un-seal" a sealed hierarchy of classes so that it can be extended.

If not, I've added some more test cases, take a look and see if your case is in there.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+}
+
+final strictfp class FilledRectangle extends Rectangle {
+}
+
+sealed interface Squircle permits Circle, Square { // ok
+}
+
+strictfp sealed interface Rhombus permits Parallelogram, Parallelogram.Diamond,
```
Comment (AkMo3) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java:37]: Please add `// violation` comment.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        Parallelogram.Trapeziod {
+
+}
+
+record Parallelogram(int x, int y, double z) implements Rhombus{
+    final public record Diamond(int x, int y, double z)implements Rhombus { // violation
+    }
+
+    strictfp public final record Trapeziod(int x, int y, double z)implements Rhombus {
```
Comment (AkMo3) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java:46]: Please add `// violation` comment

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    }
+}
+
+strictfp non-sealed interface Ellipse extends Square { // violation
```
Comment (AkMo3) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java:50]: Please remove this as this is non-compilable.

**Review by AkMo3 — CHANGES_REQUESTED** [review]
I am happy with the test cases now. Some minor:

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java (hunk 1 above)
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java:37]: Done.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java (hunk 2 above)
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java]: Done.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java (hunk 3 above)
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java]: This needed to extend `Squircle`, fixed.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java (hunk 1 above)
Comment (AkMo3) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java]: Resolved

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java (hunk 2 above)
Comment (AkMo3) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java]: Resolved

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java (hunk 3 above)
Comment (AkMo3) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java]: Resolved

**Review by AkMo3 — APPROVED** [review]

**Comment by AkMo3** [comment]
@strkkk @pbludov Please assign next reviewer

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealedNoViolation.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    private final class OtherSquare2 extends OtherSquare {
+    }
+}
+
+final strictfp class FilledRectangle extends Rectangle {
+}
+
+sealed interface Squircle permits Circle, Square { // ok
```
Comment (strkkk) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealedNoViolation.java:34]: can these classes be static? if yes, please add it to test cases

Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealedNoViolation.java:34]: Done.

**Review by strkkk — APPROVED** [review]

**Review by pbludov — APPROVED** [review]

**Comment by pbludov** [comment]
@nmancus1 JDK15 build failed: https://cloud.drone.io/checkstyle/checkstyle/2878/5/4

**Review by romani — APPROVED** [review]

**Pull request merged**